### ISSUE TITLE
produce FreeBSD assemblies for CoreFX

### ISSUE THREAD

**Issue by ghuntley** [dotnet/corefx#1626]
This issue is to discuss unit(s) of work to actually produce FreeBSD assemblies for corefx.  

> @stephentoub - There's what's likely a more pressing issue, which is actually building for FreeBSD. Today, when we need to specialize an assembly for a particular platform, we effectively have three builds, producing three different managed assemblies: Windows, Linux, OSX. Sounds like at least for now we'll need a fourth, FreeBSD. I suggest you start by modifying the build to support an IsFreeBSD property (or just IsBSD of you think there's a high chance that the implementations across BSDs will be the same even with varied kernels) along with the appropriate OSGroup targets. That can then be used in the csproj files as needed to specialize an assembly with FreeBSD-specific code.

Related issue(s) 
- #1625 (OSGroup identifier in the public API)
- #1576 (OSGroup identifier in the private API)

/cc: @janhenke @josteink

**Comment by josteink** [comment]
There seems to be agreement as far as https://github.com/dotnet/corefx/issues/1576 is concerned.

When we also have a decision on https://github.com/dotnet/corefx/issues/1625 we should be able to start shipping some code.

**Comment by ghuntley** [comment]
Agreement on #1625 has been reached by the portteam, unless MSFT chooses otherwise it will be `FreeBSD` and will be potentially introduced as part of #1999

**Comment by josteink** [comment]
> Agreement on #1625 has been reached by the portteam, unless MSFT chooses otherwise it will be FreeBSD

If I read that right, this means that when https://github.com/dotnet/corefx/pull/1999 is merged, we can consider this MSFT approving of the new public API, and can therefore press forward on the remaining issues with regular pull-requests without need for MSFT approval.

If so, that sounds good to me.

**Comment by ghuntley** [comment]
Next steps as per MSFT are https://github.com/dotnet/corefx/pull/1999#issuecomment-111279577
> 1. The "FreeBSD port team" continues their work to get a FreeBSD version of CoreFX produced (tracked by #1626).
> 2. The port team brings up enough of the CoreFX and CoreCLR stack on FreeBSD such that we can start running the CoreFX unit tests on FreeBSD.
> 3. The tests reach some minimal quality level. I don't know exactly what this looks like yet, but I expect it means something like a majority of the tests pass. Ideally we would not have a bunch of specific tests disabled for only FreeBSD (compared to Linux and OSX, we wouldn't want to hold FreeBSD to a higher standard than the other *NIX platforms we have there).
> 4. Working with the FreeBSD port team, the CoreFX team gets the CoreFX tests added to our CI system running on FreeBSD.
> 5. Discuss merging a PR based for issue #1625, which adds the property.

**Comment by josteink** [comment]
That sounds like a fully reasonable plan to me.

**Comment by janhenke** [comment]
Okay, then let's start the work on getting corefx to work.

**Comment by josteink** [comment]
First obstacle in building corefx on FreeBSD seems to be mono. The build-script insists version 4.1 is required. @ajensenwaud did some work on this on the Frankfurt-host, but I'm not sure how complete it is.

I'll queue a build for now and see what the output looks like.

**Comment by stephentoub** [comment]
> First obstacle in building corefx on FreeBSD seems to be mono

FWIW, I personally don't think this is the first obstacle. There are two build related issues:

1. Building assemblies that work correctly on FreeBSD
2. Building those assemblies on FreeBSD

(1) is critical, and is I believe what this issue is meant to be about. (2) is very nice to have, but lack of it doesn't prevent the creation of a great system for running managed code on FreeBSD.

You're of course free to prioritize however you see fit, but my recommendation would be to focus on (1) rather than (2).

Note that we barely have corefx building-on-Linux and building-on-OSX, such that our CI system builds the assemblies for those platforms on Windows; it then shuttles the resulting assemblies over to the target platform to execute the tests.

**Comment by josteink** [comment]
That's fair enough. I just assumed that it would be easier to get general FreeBSD platform support baked into corefx if we could actually build it ourselves on FreeBSD.

I'll make do with Windows-initiated building for now and attempt to ninja together a build-configuration.

**Comment by akoeplinger** [comment]
@josteink btw. corefx should now build on Mono 4.0.1.44.

**Comment by josteink** [comment]
@akoeplinger Nice. That leaves me some hope we can get it running on FreeBSD too :)

**Comment by ajensenwaud** [comment]
Good points. However if we really want corefx to be part of the FreeBSD environment, we really need it to be able to compile from source to get it into the Ports system.

I did hear that Mono 4.0.1.44 fixes a lot of these issues but have not had time to play with it yet. I know the ports team are updating the port Makefile as well as we speak with a new patch.

**Comment by stephentoub** [comment]
Yes, I'm in no way disagreeing... being able to _build_ corefx on Linux, OSX, and FreeBSD is important. I'm simply suggesting that from a priority perspective it's more important to be able to actually _run_ corefx on Linux, OSX, and FreeBSD. :wink:

**Comment by jasonwilliams200OK** [comment]
@ghuntley,
would be super :cool: if we have a markdown task checklist outlining what what is remaining:

```
- [x] task 1
- [ ] task 2
- [ ] task 3
```
renders as:

- [x] task 1
- [ ] task 2
- [ ] task 3

This will probably encourage others to score those feats and FreeBSD support will land rather sooner than anticipated! :sunglasses:

**Comment by janhenke** [comment]
To my knowledge the following pieces of work in CoreFX are required for FreeBSD support:

- [x] Introduce FreeBSD platform to the build tools and scripts. (Done by @josteink and me, PRs #2021 merged, #2030 in review)

13 Assemblies do not compile on their own and need FreeBSD specific changes. Mostly the Interop pieces that already exist for Linux/OS X (order by the occurrence in the build output):
 - [x]  ```System.Private.URI``` (done, PR #2032 merged)
 - [x]  ```System.Console``` (done, PR #2031 merged)
 - [x]  ```System.Diagnostics.Debug``` (done, PR #2039 in review)
 - [ ]  ```System.Diagnostics.Process```
 - [ ]  ```System.IO.Compression.ZipFile```
 - [ ]  ```System.IO.FileSystem.DriveInfo```
 - [ ]  ```System.IO.FileSystem.Watcher```
 - [ ]  ```System.IO.FileSystem```
 - [ ]  ```System.IO.MemoryMappedFiles```
 - [ ]  ```System.IO.Pipes```
 - [ ]  ```System.Security.Cryptography.Hashing.Algorithms```
 - [x]  ```System.Security.SecureString``` (done, PR #2039 in review)
 - [ ]  ```System.Runtime.Environment``` (blocked by #1999 )

I will try to update that list based on PRs issued and merged.

**Comment by josteink** [comment]
FYI: PR #2039 merged

**Comment by josteink** [comment]
Just trying to be ahead of the curve here... How do we plan to implement `System.IO.FileSystem.Watcher` ?

Iirc FreeBSD has no `inotify` such as Linux and Windows does (which is also why there is no Dropbox last time I checked). Will this be a potential source of trouble coming our way? Or does anyone have an idea for how to work around this?

**Comment by janhenke** [comment]
I suggest we stub that out for the moment and throw a PlatformNotSupportedException as Stephen Toub suggested in the other topic (https://github.com/dotnet/corefx/pull/2021#issuecomment-111602342). Then we have at least a complete set of assemblies and we can continue to work on that particular issue without blocking further steps.

Would you mind opening a separate issue for that?

**Comment by ghuntley** [comment]
Let's move `System.IO.FileSystem.Watcher` discussions to #2046

**Comment by jasonwilliams200OK** [comment]
Guys is there any such blocker for `System.Diagnostics.Process`?

**Comment by ghuntley** [comment]
FreeBSD was merged into S.RT.I.RI early this morning by @jasonwilliams200OK , unit tests need to be added once build tools are added:

* [ ] Update build tools to support FreeBSD - https://github.com/dotnet/buildtools/pull/180
* [ ] Reapply `[Fact,PlatformSpecific(PlatformID.FreeBSD)]` tests to CheckPlatformTests  - see https://github.com/dotnet/corefx/pull/2068#discussion_r32570244 and https://github.com/dotnet/corefx/pull/2068#discussion_r32570244

**Comment by ghuntley** [comment]
@jasonwilliams200OK there were some discussions last night about `System.Diagnostics.Process` in gitter which have been formalized into https://github.com/dotnet/corefx/issues/2070

**Comment by jasonwilliams200OK** [comment]
@ghuntley, thanks. I actually read those messages. `System.Diagnostics.Process` is a tricky one. AFAIK, io.js team had similar challenges with FreeBSD process management. Mono team has probably nailed it, so lets hope if @akoeplinger and co. could enlighten us on this matter? :)

**Comment by josteink** [comment]
>  System.IO.FileSystem.DriveInfo

As discussed in the gitter, For this one I tried looking into basic usage of `getmntinfo`:

````c
#include <sys/param.h>
#include <sys/ucred.h>
#include <sys/mount.h>
#include <stdio.h>

int main() {
  struct statfs *mntbuf;
  int mntsize = getmntinfo(&mntbuf, MNT_NOWAIT);

  for( int i = 0; i < mntsize; i++ ) {
    printf("%s\n", mntbuf[i].f_mntonname);
  }
}
````

Running that sample yielded this output:

````
$ ./a.out
/
/dev
/tmp
/usr/home
/usr/ports
/usr/src
/var/crash
/var/log
/var/mail
/var/tmp
/dev/fd
/usr/compat/linux/proc
/proc
$
````

So it seems it does what we need. The question is, should we do any type of filtering on the results?

Looking at the "intent" of the `DriveInfo` object, coming from the Windows world of .NET it has often been to enumerate the available locations to store or retrieve files (`C:`, `D:`, etc). But when using Unix hierarchical file-systems, returning `/` would be adequate to cover those needs.

So what should we return? What would be useful? Should even consider it being useful or not?

The Linux-version just dumps everything, except things set to be ignored:

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.DriveInfo/src/System/IO/DriveInfo.Linux.cs#L98-L99

I tried putting in the following filter, but it didn't really change anything in terms of output:

````c
    if ((mntbuf[i].f_type != MNT_IGNORE)) {
        printf("%s\n", mntbuf[i].f_mntonname);
    }
````

Any opinions?

**Comment by jasonwilliams200OK** [comment]
@josteink, great diggings! Based on https://github.com/dotnet/corefx/issues/815#issuecomment-113825960 and https://github.com/dotnet/corefx/issues/1729, I think we should collaborate with @sokket to come up with a solution with works  across different Unices.

**Comment by sokket** [comment]
I have a version running on OSX that uses getmntinfo and statfs to get information about each mount point, which seems like the most logical mapping from the Windows Drive concept. I'll double check that the function and struct definitions on OSX match the FreeBSD definitions and, if so, my commit for OSX will work for BSD as well.

I'll be sure to add you to my PR @josteink

**Comment by josteink** [comment]
Sounds good. Thanks for the heads up and thanks for giving FreeBSD  some love too.

I looked into some basic pinvoke for these functions, and it seems like we need to do all the marshalling and conversions ourselves, so if anyone else has already put in the effort, who am I to say no? ;)

**Comment by sokket** [comment]
No problem...looks like the main difference was with the struct declarations; since we'll probably hit this more in the future, I'm doing some refactoring that will allow us to share a lot of the PInvoke signatures. I'll add a bigger description in my PR (today or tomorrow, based on how the test run) but I basically added the PInvoke signatures and struct signatures for FreeBSD (based on the headers I found online) and it compiles. I've tested it on Mac so it *should* (in theory...) work on FreeBSD since it's just a struct declaration change, but your milage may vary :). If it doesn't, you'll have the DriveInfo class and PInvokes 99% of the way there and will just require some tweaking based on FreeBSD nuances.

**Comment by ghuntley** [comment]
Excellent news @sokket. I've created you an account on the machine the port-team uses for development, it's european based but it's always on and has heaps of memory and processing power. 

    ssh sokket@freebsd-frankfurt.zbsd.org

Password authentication is disabled, use [one of your keys](https://api.github.com/users/sokket/keys).

See also issue: https://github.com/dotnet/corefx/issues/815 (System.IO.FileSystem.DriveInfo for Mac/FreeBSD)

**Comment by janhenke** [comment]
Are there any updates? Did anybody implement the remaining assemblies on FreeBSD?

**Comment by josteink** [comment]
I've been busy attending my new baby, haven't had time for any coding anywhere.

I've suspected that issues like these have been lying dormant and I guess this confirms it to a certain extent.

**Comment by janhenke** [comment]
For the assemblies that are still not implemented, I linked the "how to implement"-issue in the list above. I hope that helps coordinating the effort to get these remaining assemblies implemented.

**Comment by josteink** [comment]
I must admit I was having difficulties keeping track of what we have done and where, so that's definitely a good move. Good job :)

**Comment by ajensenwaud** [comment]
Where do I find that? Would be grat to get the remaining assemblies 
implemented.

**Comment by josteink** [comment]
This comment here: https://github.com/dotnet/corefx/issues/1626#issuecomment-111800540

**Comment by janhenke** [comment]
I am right now waiting for the native shims to be finished, as these should take over most of the work for getting these assemblies working on FreeBSD.

@nguerrera would be great if you can keep us posted on the progress. :)

**Comment by jasonwilliams200OK** [comment]
Update:
@janhenke confirmed that with https://github.com/dotnet/corefx/pull/2974 merged, `System.IO.Pipes` builds on FreeBSD! :sunglasses:

**Comment by jasonwilliams200OK** [comment]
Update:
#2527 closed, ` System.IO.MemoryMappedFiles ` builds on FreeBSD.
Thanks @janhenke for the confirmation!

**Comment by janhenke** [comment]
Thanks to the shims approach, it just comes down to make sure the shims compile on FreeBSD. Thankfully that makes life a lot easier. :)

**Comment by janhenke** [comment]
#3257 should bring us both ```System.Diagnostic.Process``` and ```System.IO.FileSystem.Watcher``` leaving just ```System.Net.NameResolution``` unresolved. (I will check the mentioned two assemblies once the PR is merged and works on FreeBSD)

**Comment by janhenke** [comment]
#3471 should bring us ```System.Net.NameResolution``` and complete the list above.

**Comment by sokket** [comment]
#3471 was just merged :)

**Comment by jasonwilliams200OK** [comment]
@sokket, thanks for the update. I built master (f467911) on FreeBSD using this guide: https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24. Current blocker is https://github.com/dotnet/buildtools/issues/292, which is fixed in upstream but waiting for next buildtools roll-out. :)

**Comment by jasonwilliams200OK** [comment]
Update: new buildtools with fix for dotnet/buildtools#292 has landed in CoreFX master. Next stopper from buildtools is https://github.com/dotnet/buildtools/issues/300: missing OS specific tool to be able to run the tests.

@janhenke, you have marked `System.Diagnostics.Process` (#2070) and `System.IO.FileSystem.Watcher` (#2046) as done; but they are neither implemented nor do they compile on FreeBSD. Have you actually verified the list by compiling the managed code?

Based on my recent experience with commit 60c78da3c918b0d256cc1f878de06d351dbe3342 (see [msbuild.log](https://gist.github.com/jasonwilliams200OK/0a2e92e19eb5bb5f3130)), following assemblies do not compile:

* System.Diagnostics.Process
* System.Diagnostics.ProcessManager
* System.Diagnostics.ThreadInfo
* System.IO.FileSystemWatcher
* System.Net.SocketAddress *(alright, this one was added recently)*

**Comment by janhenke** [comment]
As far as I recall I verified the associated shims compile. Since the managed code should be free of FreeBSD specific code. Those shims you mention should have been shimed out with the PRs linked above.

**Comment by jasonwilliams200OK** [comment]
>  Those shims you mention should have been shimed out with the PRs linked above.

Actually, PR https://github.com/dotnet/corefx/pull/3257 is not related to shim. There is still some PAL code within the managed projects (the old approach), therefore it is required to build managed assemblies to be absolutely sure.

**Comment by janhenke** [comment]
> Actually, PR #3257 is not related to shim.

I disagree. It is refactoring the P/Invoke code to the System.Native shim. Also as I edited above, I am recalling at least some of the assemblies compiled in between.

**Comment by jasonwilliams200OK** [comment]
> I disagree

https://github.com/dotnet/corefx/pull/3257/files: see the instances of `.Unix.cs` and `.Linux.cs` for `System.Diagnostics.` and `.OSX.cs` are untouched.

> It is refactoring the P/Invoke code to the System.Native shim

Yes it does refactor some common helpers under `System.Native`, but not for `System.Diagnostics.*` et al.

**Comment by stephentoub** [comment]
Even when these assemblies are only P/Invoking to System.* libs, there may still be FreeBSD work required for some of them, e.g. System.Diagnostics.Process and System.IO.FileSystem.Watcher. They are using functionality specific to Linux and OS X, and we don't plan to try to abstract that behind native shims. The goal of the shims isn't to end up with a single managed binary for Unix, though that's a very nice property when it comes from the work; the primary goal is to avoid ABI differences that cause fragility. I expect at least a handful of assemblies will continue to have Linux/OS X specific binaries, where a FreeBSD binary would also be needed.

**Comment by stephentoub** [comment]
FYI, there are no corefx assemblies named System.Diagnostics.ProcessManager, 
System.Diagnostics.ThreadInfo, 
System.IO.FileSystemWatcher, or
System.Net.SocketAddress. Those are types in other assemblies.

**Comment by jasonwilliams200OK** [comment]
> I expect at least a handful of assemblies will continue to have Linux/OS X specific binaries, where a FreeBSD binary would also be needed.

Does that mean whenever Solaris and non-glibc (musl and  μlibc) targeting Linux such as Alpine supports will arrive, they will have separate binaries? And then different architectures ARM, MIPS, RISC, SPARC etc. would require another level of separation?

Wouldn't it make sense to converge them to POSIX interface / sys-calls as much as possible and feature-detect the differences using configs (via CMake) to be used in the same binary (unless it is impacting the size/performance of the assemblies significantly)? As I have understood it, `System.Native.so` binary has common helper for other specific `System.*.Native.so` which seems enough for separation-of-concerns principle compliance. But if it gets transformed to `System.Net.Http.FreeBSD.ARM.Native.so` or `System.Net.Http.Solaris.SPARC.so`, then it will be quite unmanageable with "too many moving parts" etc.

> there are no corefx assemblies named

Good point. I was actually going by the failures instances in msbuild logs and number of `.OSX.cs` and `.Linux.cs` files. :smile:

**Comment by stephentoub** [comment]
> Wouldn't it make sense to converge them to POSIX interface / sys-calls as much as possible

We do. How do you propose doing file watching well via POSIX?

**Comment by jasonwilliams200OK** [comment]
> How do you propose doing file watching well via POSIX?

We cannot do all of it POSIX, since inotify is Linux specific. FreeBSD/OSX offers a separate implementations.

#### Proposal:

From native binaries distribution viewpoint, every OS should receive equal set of binaries with same names, but different functionality.

Here is a proposed structure:

```cmake
# cmake

check_include_files( "inotify.h" HAVE_INOTIFY_ABILITY )
```

---

```inf
// config.h.in
cmakedefine01 COREFX_HAVE_INOTIFY_ABILITY
// always a good idea to prefix our headers with project id :)
```

---

```c++
// header (pal_fsw.hpp) file

#pragma once

class file_system_watcher_shim
{
public:
  void common_function_for_posix_compliants();
  void slightly_diverged_function();
  void painfully_diverged_watch_function();
}
```
---

```c++
// source (pal_fsw_commons.cpp) file

#include "pal_fsw.hpp"

void file_system_watcher_shim::common_function_for_posix_compliants() {
 // TODO: implement common function for all
}

void file_system_watcher_shim::slightly_varied_function() {

#if COREFX_HAVE_XYZ_ABILITY

  // your way

#else

  // my way

#endif // COREFX_HAVE_XYZ_ABILITY

}
```

---

```c++
// source (pal_fsw_inotify.cpp) file

// this is a separate compilation unit and will clash with others,
// therefore guarding it with preprocessor directive

#if COREFX_HAVE_INOTIFY_ABILITY

#include "pal_fsw.hpp"

void file_system_watcher_shim::painfully_diverged_watch_function() {
 // TODO: implement inotify based watcher
}

#endif // COREFX_HAVE_INOTIFY_ABILITY
```
---

```c++
// source (pal_fsw_non_inotify.cpp) file

// this is a separate compilation unit and will clash with others,
// therefore guarding it with preprocessor directive

#if !COREFX_HAVE_INOTIFY_ABILITY

#include "pal_fsw.hpp"

void file_system_watcher_shim::painfully_diverged_watch_function() {
 // TODO: implement non-inotify way
}

#endif // !COREFX_HAVE_INOTIFY_ABILITY
```

**Comment by stephentoub** [comment]
This is essentially what we have, e.g.
- "common_function_for_posix_compliants" are natively-shimmed 1:1 functions consumed from logic in a shared Unix managed binary
- "slightly_diverged_function" are natively-shimmed almost 1:1 functions with some native #ifdefs consumed from logic in a shared Unix managed binary
- "painfully_diverged_watch_function" are / will be natively-shimmed 1:1 functions consumed from logic in platform-specific managed binaries

The real difference is whether the code implementing the "painfully diverged" logic is done in C# or C++, and we've chosen C# and it's already all implemented in C#.  I've not seen any compelling argument for why in such cases rewriting it all to be in C++ would be a significantly more compelling option.

**Comment by janhenke** [comment]
@jasonwilliams200OK With today's update to mono I build corefx on FreeBSD itself again. There are a lot of new error messages since the last time I build it.
I am wondering if ```Interop.Libc``` will go away eventually?

@stephentoub It all comes down to packaging. Having all platform specific code in the native part has the benefit of having one managed assembly for all Unix-like platforms.
Besides, I strongly think we need a generic implementation for these "platform dependent managed code. Even if it just throws a NotImplementedExcpetion. That way you can much easier port to new platforms, if it at least compiles everything right away. Also it would give the chance to at least try running on an unsupported platform.
Generally it is much easier if you can at least compile successfully right away.

**Comment by jasonwilliams200OK** [comment]
@stephentoub, sorry I was mixing C++ with C#. Now i understand that native layer is just exposing those entry points (native libs funcitons or syscalls) and sanitizing IO and managed code is where we decide which platform-specific wrapped utility method / wrapped syscall to be consumed. Besides, both (native and managed) tiers can't be OS-agnostics where non-POSIX functionality is to be consumed.

@janhenke, I agree. I am also building master as we speak. There is one recurring assembly signing issue:
```
CSC : error CS7027: Error signing output with public key from file '/root/corefx/packages/Microsoft.DotNet.BuildTools.1.
0.25-prerelease-00104/lib/ECMA.snk' -- mscoree.dll [/root/corefx/src/System.IO.Compression.ZipFile/ref/System.IO.Compres
sion.ZipFile.csproj]
CSC : error CS7033: Delay signing was specified and requires a public key, but no public key was specified [/root/corefx
/src/System.IO.Compression.ZipFile/ref/System.IO.Compression.ZipFile.csproj]

CSC : error CS7027: Error signing output with public key from file '/root/corefx/packages/Microsoft.DotNet.BuildTools.1.
0.25-prerelease-00104/lib/ECMA.snk' -- mscoree.dll [/root/corefx/src/System.IO.Compression/ref/System.IO.Compression.csp
roj]
CSC : error CS7033: Delay signing was specified and requires a public key, but no public key was specified [/root/corefx
/src/System.IO.Compression/ref/System.IO.Compression.csproj]
```

will post the full msbuild.log gist link shortly.

> Besides, I strongly think we need a generic implementation for these "platform dependent managed code. 

I agree. Instead of partial classes, we can probably use inheritance with common and mostly common virtual methods in abstract base class and override for "mostly common / partly different" where necessary. Then implement abstract methods which are totally different for each OS. With this approach IMO, if where specialization/generalization is losing DRY'ing, we can employee multi-degree inheritance ancestry. But last time I checked, partial classes were preferred over parent-child association in CoreFX (for some reason I don't recall). :)

**Comment by jasonwilliams200OK** [comment]
Here is the msbuild.log for commit 2d36889:

https://gist.githubusercontent.com/jasonwilliams200OK/18c846d67ee8f4b9662e/raw/39ec6f9a4d9da7579935a0c259ec23d59dc99917/corefx-msbuild.log

/cc: @akoeplinger

**Comment by janhenke** [comment]
@jasonwilliams200OK Why so complicated. All it needs is a "if not Windows,Linux,OS X" condition in the project files. So either include a platform specific set of files in the build or the generic ones.

**Comment by nguerrera** [comment]
I don't think the fact that some assemblies can't build/work for FreeBSD should be a major blocker for testing the rest of them. We should probably just make it so that the ones with pending work (FileSystemWatcher, Process, Networking) are skipped in the build and test run on FreeBSD. Then we can bring those up individually while having a process to prevent regression in what already works.

**Comment by stephentoub** [comment]
> I don't think the fact that some assemblies can't build/work for FreeBSD yet should be a major blocker for testing the rest of them

Agreed

> We should probably just make it so that the ones with pending work (FileSystemWatcher, Process, Networking) are skipped in the build and test run on FreeBSD

Or similar to what @janhenke suggested, just allow them to build, either by stubbing out with files that throw PlatformNotSupported, or just by mapping FreeBSD to one of the cases that does work, e.g. if FreeBSD is chosen, just build the Linux one (it won't work, but it'll build).

**Comment by jasonwilliams200OK** [comment]
With @ellismg recent changes (#3684), I am able to run tests by simplifying the previous method (https://github.com/dotnet/coreclr/issues/1633#issuecomment-143669303):

* After https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24 (especially the step to build native shims separately *after* the build.sh exit with status 1), download [CoreCLR artifacts](http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/) zip: `cd /root; curl -o bins.zip "http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/lastSuccessfulBuild/artifact/*zip*/archive.zip"` (don't forget the quotes around URL) and then `unzip bins.zip; chmod -R 0777 archive/; rm bins.zip; cd corefx`.

(nothing required from Windows machine)

* Ran the test:

  ```
  ./run-test.sh \
  --coreclr-bins      ../archive/bin/Product/FreeBSD.x64.Debug \
  --mscorlib-bins      ./packages/Microsoft.DotNet.CoreCLR/1.0.0-prerelease/lib/aspnetcore50 \
  --corefx-native-bins ./bin/FreeBSD.x64.Debug/Native
  ```

It says:

>
40 test(s) failed

**Comment by josteink** [comment]
> I don't think the fact that some assemblies can't build/work for FreeBSD yet should be a major blocker for testing the rest of them.

I have to agree with this one. It's better to start QAing the work we have done, instead of waiting on everything to complete before starting testing.

**Comment by jasonwilliams200OK** [comment]
> 40 out of how many? In what ballpark are we? :)

beats me too. The tests are spawning out of test assemblies (managed dlls) and total number of tests is not visible.

**Comment by ellismg** [comment]
The number the script produces at the end is the number of test assemblies that failed.  xUnit should write the number of tests that failed per assembly to stdout as part of it's run.  The numbers should also be in the XML files it produces in each test assembly folder.

It could be the runtime is also just crashing and in that case there may not be logs produced per test assembly.

**Comment by naamunds** [comment]
@jasonwilliams200OK Do you know if any progress has been made on the assembly signing issue?  I'm hitting the same thing on Ubuntu.  If no one's working on it, perhaps we should open a separate issue for it.

**Comment by jasonwilliams200OK** [comment]
@naamunds, that has been fixed on CoreFX master as part of https://github.com/dotnet/corefx/issues/3739.

**Comment by jasonwilliams200OK** [comment]
Update - Today I ran test on FreeBSD, thousands of tests were passing and then some were failing due to obvious lack of System.Diagnostics.Process and friends snafu. After ~40 minutes of successful execution, it hung on System.IO.FileSystem tests (for about ~15 minutes before I pressed Ctrl+C to terminate).

**Comment by sec** [comment]
@jasonwilliams200OK how did you manage to compile corefx under freebsd? I'm stuck at errors about gssapi

**Comment by jasonwilliams200OK** [comment]
@sec, at the time of making these notes: https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24, GSSAPI wasn't required by CoreFX. However, it seems like the pkg is recently ported to FreeBSD: http://www.freshports.org/security/p5-GSSAPI. You may want to try `pkg upgrade pkg && pkg update && pkg install p5-GSSAPI`.

**Comment by sec** [comment]
@jasonwilliams200OK, I already got this (it's perl ext. btw.) Problem was missing gssapi_ext.h. The trick was to do "pkg install krb5" - now native compiled.
I've copied them to coreclr runtime, but still can't run ASP.NET Core app :) fight goes on then.

**Comment by josteink** [comment]
What's the current status of this task? Is @janhenke's list complete and accurate? Is all work which needs being done, done? Then it should be closed, right?

If so, why do we still have this task? https://github.com/dotnet/corefx/issues/2070

If there's still work to be done, should a new issue be registered based on the current state of affairs?

**Comment by sec** [comment]
There's also this needed I think - #2046

**Comment by jasonwilliams200OK** [comment]
> should a new issue be registered based on the current state of affairs?

Yes :+1:

**Comment by karelz** [comment]
We discussed community-driven port for FreeBSD with @RussellHaley (from FreeBSD community) and @wfurt (from .NET Core team) who both expressed interest in the work.
Here's a plan proposal we put together (feedback / suggestions are welcome):

1. Produce binaries in CoreCLR & CoreFX repo targeting FreeBSD - using hacks is fine
    * Hard to parallelize, @wfurt will work on that
    * The build can be mix of builds from other platforms (Mac, Linux) targeting FreeBSD
    * We will need documented steps (on FreeBSD wiki) to reproduce the build with FreeBSD-specific bug fixes
2. Run & stabilize CoreCLR tests (using corerun)
    * Tests may be built on another platform
    * Goal: Provides basic quality of runtime
3. Run & stabilize CoreFX tests (using corerun)
    * Tests may be built on another platform
    * Note this requires xunit. We believe, based on our past porting experience, once [2] is done, xunit will just work.
    * This can be in theory parallelized with [2] - it may require shortcutting xunit (e.g. generate static execution recipe on another platform)
4. Full stack build on FreeBSD (using corerun as bootstrapper from [1]-[3])
    * We will need all tools (nuget, msbuild, roslyn) to work on boostrapping .NET Core
5. Installers (FreeBSD ports)
    * First-stage: Using product binaries from nuget feeds
    * Second-stage: Build product from source (blocked on build from source effort)
    * Requires FreeBSD community expertise and guidance on design
    * Note: We can link FreeBSD packages also from official .NET Core download pages as community-support packages
6. Regular build and test runs on FreeBSD
    * Goal: Make sure changes in .NET Core repos breaking FreeBSD are known early
    * Design needed
    * Requires FreeBSD community expertise and guidance on design

Operation principles:
* Changes in [2]-[4] should be done primarily in CoreCLR/CoreFX repos (due to CLA signing requirements, code reviews from .NET Core team experts/members. etc.)
* We will track high-level work on this issue. Specific bugs will be filed as separate issues.

If anyone is interested in helping, please let us know here. We can easily distribute work items from [2] & [3] above once we are far enough with [1].

**Comment by karelz** [comment]
Tagging more folks who expressed interest on freebsd-mono list ([this thread](https://lists.freebsd.org/pipermail/freebsd-mono/2017-September/002558.html)): @smortex @radovanovic

BTW: I can't find [Mathieu Prevot](https://lists.freebsd.org/pipermail/freebsd-mono/2017-September/002570.html) GitHub account.

**Comment by krytarowski** [comment]
For NetBSD we miss robust posix mutexes, this is the only feature that is still missing to produce named robus mutexes. We can now debug managed code failures with LLDB/NetBSD.. it works fine with core files. In my previous attempts we died on lack of this feature in LLDB (I started to port this debugger for .NET).

Having better FreeBSD support will significantly help.

There were also problems in the past with lack of hyperv guest support to attach a NetBSD buildbot to CI machines and verify patches... for this I might need help from MS. I expect that there is required proprietary software to run it, which I don't own... I might find a developer to do the job if there would be joint interest (investment) between The NetBSD Foundation and Microsoft.

**Comment by karelz** [comment]
Where do we miss "robust posix mutexes"? Is it part of .NET Core PAL?

Which CI system are you referring to? Why is it tied to .NET Core port effort?

**Comment by krytarowski** [comment]
> Where do we miss "robust posix mutexes"?

In the NetBSD kernel (and libc/libpthread), this is a part of CoreCLR. FreeBSD developed it in the last two years. It might be available in the latest stable release.. but there is need to check.

I want to add this feature before my restart of .NET porting. (There was also detected a tiny missing API for network routing.. but I skip it now).

> Is it part of .NET Core PAL?

I don't remember now, without checking the code - it's the API used to implement of .NET named robust mutexes (or perhaps semapthores).

> Which CI system are you referring to?

NetBSD.

> Why is it tied to .NET Core port effort?

It was optional feature last time I looked. I decided to get feature-parity on the kernel-interfaces and utilities (like LLDB). Just my style of work, to get functional buildingblock and later build a house. At some point we will need it anyway so why not to develop it in one go. Thanks for understanding :)

**Comment by RussellHaley** [comment]
Perhaps you can just tag the freebsd-dotnet group on GH? He is a member there (or you could look up his account name). ‎His email is mprevot@freebsd.orgRuss                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone on the Virgin Mobile network.                                                                                                                                                                                                                From: Karel ZikmundSent: Friday, September 15, 2017 10:52 AMTo: dotnet/corefxReply To: dotnet/corefxCc: Russell Haley; MentionSubject: Re: [dotnet/corefx] Support for FreeBSD (#1626)Tagging more folks who expressed interest on freebsd-mono list (this thread): @smortex @radovanovic
BTW: I can't find Mathieu Prevot GitHub account.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/dotnet/corefx","title":"dotnet/corefx","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/dotnet/corefx"}},"updates":{"snippets":[{"icon":"PERSON","message":"@karelz in #1626: Tagging more folks who expressed interest on freebsd-mono list ([this thread](https://lists.freebsd.org/pipermail/freebsd-mono/2017-September/002558.html)): @smortex @radovanovic\r\n\r\nBTW: I can't find [Mathieu Prevot](https://lists.freebsd.org/pipermail/freebsd-mono/2017-September/002570.html) GitHub account."}],"action":{"name":"View Issue","url":"https://github.com/dotnet/corefx/issues/1626#issuecomment-329853313"}}}

**Comment by karelz** [comment]
@RussellHaley feel free to tag the larger group if you think it is appropriate. I couldn't find Mathieu's GH account via his name nor email, that's what I meant above.

**Comment by RussellHaley** [comment]
I will look at tagging the group.

Here is Mathieu's account. Perhaps a privacy setting?

https://github.com/mprevot

Cheers,

Russ

**Comment by radovanovic** [comment]
I can't see anywhere here mentioned, but what lowest version of FreeBSD we are targeting here (I assume at least 10 and later, but maybe 9 as well)?
(I am also little bit confused what discussion should happen on mono@freebsd mailing list, and what here?)

**Comment by Echo-8-ERA** [comment]
Well, if Fedora is anything to go by, only 10.3 (10.4 soon) and 11.1.

**Comment by sec** [comment]
@radovanovic FreeBSD 9 is not supported anymore, 10 will EoL in april 2018, 11 in 2021. From my expierence there shouldn't be any problems with compiling on 11 vs 10 (and even 9 if you want). FreeBSD is developed with backward compatibility in mind.

**Comment by karelz** [comment]
> @radovanovic I am also little bit confused what discussion should happen on mono@freebsd mailing list, and what here?

I expected the technical discussions, coordination of work and status over here as this is wider audience than mono@freebsd mailing list. OTOH we don't want to have gazillion of random discussions on one issue, so we might take some specific design discussions into separate issues from this one if they grow above reasonable number of replies.

**Comment by wfurt** [comment]
I was finally able to run corefx tests on FreeBSD 11.0 (without outerloop tests)  
Total passed: 144208
Total failed: 2622
Total skipped 207

I will update https://github.com/dotnet/corefx/wiki/Building-.NET-Core--2.x-on-FreeBSD with instructions.  I will file specific issues and tag them with with os-freebsd and up-for-grab.
Full scale battle can start. Volunteers needed.  

And yes, I did skip proposed step two. I'll get back to it as well.

**Comment by wfurt** [comment]
With some work-in-progress changes the current stats look like:
Total passed: 238892
Total failed: 58
Total skipped 1628

System.Runtime.Tests.dll, 1
System.Net.Ping.Functional.Tests.dll, 7
System.Net.NameResolution.Pal.Tests.dll, 3
System.Net.NameResolution.Functional.Tests.dll, 4
System.IO.MemoryMappedFiles.Tests.dll, 1
System.IO.FileSystem.Tests.dll, 7
System.Globalization.Tests.dll, 2
System.Drawing.Common.Tests.dll, 31
System.Console.Tests.dll, 2

#24538 opened to track broken cups support.

**Comment by krytarowski** [comment]
Great progress! Adding NetBSD when having FreeBSD support in-tree should be simple.

**Comment by krytarowski** [comment]
@wfurt please share E-mail address, I will drop few lines.

**Comment by krytarowski** [comment]
Can we already bootstrap .NET on FreeBSD?

**Comment by wfurt** [comment]
I have not tried for a while @krytarowski There was push to update tooling to 2.0 release and I was waiting for that effort to stabilize. I'll give it another try and I'll post update.

**Comment by RussellHaley** [comment]
Hi, so I'm bogged down with the clr managed tests not running. https://pastebin.com/B5KhtKX5

Any input would be great as that's been an issue for some time. I have also recently run into a build failure on corefx building on Windows(master, Git revision 749194e). https://pastebin.com/JXUySLTY

I assume that's an intermittent problem but it's slowed me down tonight.

**Comment by josteink** [comment]
If you look at the error:

````
tests/runtest.sh: line 786: ((: i<: syntax error: operand expected (error token is "<")
````

And the [offending line of code](https://github.com/dotnet/coreclr/blob/master/tests/runtest.sh#L786):

````bash
for (( i=0; i<$maxProcesses; i++ )); do
````

My gut feeling would be that `$maxProcesses` is not defined, leading to an incomplete boolean expression:

````diff
+for (( i=0; i<$maxProcesses; i++ )); do
-for (( i=0; i<; i++ )); do
````

This should be fairly testable. And if that is the case, you just have to go hunting backwards to try to find out how you ended up like this.

**Comment by RussellHaley** [comment]
Thanks for the push @josteink , I was being lazy. :) You're correct. 
/*************************************/
diff --git a/tests/runtest.sh b/tests/runtest.sh
index 12b0e6d74..5c45fc74f 100755
--- a/tests/runtest.sh
+++ b/tests/runtest.sh
@@ -113,6 +113,10 @@ case $OSName in
         libExtension="so"
         ;;

+    FreeBSD)
+        libExtension="so"
+        ;;
+
     NetBSD)
         libExtension="so"
         ;;
@@ -736,7 +740,9 @@ function run_test {
 }

 # Variables for running tests in the background
-if [ `uname` = "NetBSD" ]; then
+if [ `uname` = "FreeBSD" ]; then
+    NumProc=$(getconf NPROCESSORS_ONLN)
+elif [ `uname` = "NetBSD" ]; then
     NumProc=$(getconf NPROCESSORS_ONLN)
 elif [ `uname` = "Darwin" ]; then
     NumProc=$(getconf _NPROCESSORS_ONLN)
@@ -749,6 +755,7 @@ else
         NumProc=1
     fi
 fi
+
 ((maxProcesses = $NumProc * 3 / 2)) # long tests delay process creation, use a few more processors

 ((processCount = 0))
@@ -783,6 +790,7 @@ function waitany {
 function get_available_process_index {
     local pid
     local i=0
+
     for (( i=0; i<$maxProcesses; i++ )); do
         pid=${processIds[$i]}
         if [ -z "$pid" ] || [ "$pid" == "$pidNone" ]; then

/***************************************************************/
This allows the tests to run, but I gave up at ~1000 errors all of the same nature:

FAILED   - JIT/Methodical/casts/iface/_il_dbgiface2/_il_dbgiface2.sh
               BEGIN EXECUTION
               /usr/home/russellh/Git/coreclr/bin/tests/Windows_NT.x64.Debug/Tests/coreoverlay/corerun _il_dbgiface2.exe
               coreclr_initialize failed - status: 0x80004005
               Expected: 100
               Actual: 255
               END EXECUTION - FAILED

**Comment by RussellHaley** [comment]
Okay, as per the very excellent information from janvorli, I was running part of my build in release and part of my build in debug. An embarrassing copy/paste mistake. I'm rebuilding now.

https://github.com/dotnet/coreclr/issues/1419

**Comment by josteink** [comment]
> Patch is here: https://pastebin.com/d5y9k1tw

If you have a patch which fixes a broken build, I would recommend sending it as a pull-request so that it gets fixed for everyone :)

**Comment by RussellHaley** [comment]
Thank you, I will. I'm still working on getting tests to run though and I wasn't sure where the subsequent error was last night.

**Comment by oliverw** [comment]
I suppose Freebsd 11 "pkg install" support for netcore 2.1 (once released) won't happen, right?

**Comment by RussellHaley** [comment]
@wfurt was able to get the CLR and the FX to build using Linux but it was largely untested. I was able to get the 'native' parts to build on FreeBSD but stalled out getting the managed parts to build on Windows (for FreeBSD). The whole thing was a mess of transferring files between operating systems.

Separately on the freebsd-mono@freebsd.org mailing list, @dragonsa has imported a binary version of Dot Net Core 1 and all the toolschain (msbuild, nuget etc) from MintOS using Linux emulation. I've been able to use his patches and run some of the tools but never got around to building anything. I'm not sure if these patches have been committed yet; I was in the middle of reviewing them and I changed jobs. I don't have any DotNet in my current role and am ramping on other things right now.

What does all that mean? If someone can push @dragonsa's patches into ports/pkg, then it's technically possible to build core 2 on FreeBSD natively. As you can see, though, there are lots of little parts that need to be brought together and organized. I've dropped the ball on that so someone needs to pick it up. I suggest jumping on the freebsd-mono@freebsd.org mailing list. https://www.freebsd.org/doc/en_US.ISO8859-1/books/handbook/eresources-mail.html

**Comment by oliverw** [comment]
@russellhadley Thanks for the writeup Russell.

**Comment by cschuber** [comment]
Hi,

Discussing this with a .NET dev here, I'm willing to assist with the development of a FreeBSD port/package.

Full disclosure: I'm not a .NET developer however I'm willing to work with whomever to get this into the tree.

~cy

**Comment by josteink** [comment]
@cschuber I've been too busy to keep an eye on the current status of things, but as someone who submitted a lot of FreeBSD-patches and managed to get Hello World running around 3 years ago, it would be awesome if we could *finally* see this thing getting properly landed. You have my full support :)

**Comment by kasper3** [comment]
@cschuber, the current active issues is https://github.com/dotnet/coreclr/issues/18067. Mainly these four features are left https://github.com/dotnet/corefx/issues?q=is:open+label:os-freebsd+label:up-for-grabs+is:issue, among which Filesystem watcher seems to be the most tricky/laborious one https://github.com/dotnet/corefx/issues/2046.

**Comment by wfurt** [comment]
Thanks for the offer @cschuber. We are almost at point when it may be possible. 
We've been working recently with @mateusrodrigues on getting .net working on FreeBSD and he is trying to get PowerShell working. List @kasper3 sent are are primarily missing features. I think we can throw PNSP for now. From my prospective most pressing issues are #30698 and https://github.com/dotnet/coreclr/issues/18481.  It would be great if anybody from the community can dig into them. I did not run tests recently, but I fear the number of failures went up.
We should open issue for each new failing group. 

I also start fixing source-build but there are still some challenges ahead. 
c# compiler is written in c#. Current .net build uses previous version of .net to produce managed assemblies. It also depends on packages from Nuget. 
Right now, we have good enough bootstrap cli that we can be build coreclr, corefx and few other repos on FreeBSD. I did not update building instructions yet to reflect 2.1 changes and source-build.

**Comment by josh-endries** [comment]
+1 Just a note to say I'm glad this still has some momentum. It's hard to follow with so many moving parts but seems like people are making progress. I created https://github.com/dotnet/coreclr/issues/6115 a while ago but the project I was working on then got put on hold. I really hope it's as easy as `pkg install dotnet && dotnet build` one day (without linux compat).

**Comment by bott0r** [comment]
Also looking forward for this

**Comment by stephentoub** [comment]
Nice job, @wfurt and @bartonjs.

**Comment by josteink** [comment]
When I proposed my first FreeBSD commits around 2-3 years ago, I didn’t actually *believe* we’d get here, but I certainly wanted to try. 

This is definitely a big milestone and hopefully will make it easier for new contributors help finish the port. 

Big thanks to everyone who helped us get this far 👍

**Comment by krytarowski** [comment]
Great progress! I'm still fighting with toolchain (most LLVM projects besides LLDB and LLD are finished) and hardware assisted virtualization for NetBSD (Linux/BSD now starts booting until a VTx fatal exception, simpler OSes like FreeDOS already work).. so I will resume my porting, hopefully sooner than later. Better FreeBSD support will help with an easier resume.

**Comment by bott0r** [comment]
Awesome :)

**Comment by peturingi** [comment]
We celwbrate drunkness why the ensil bombsrdment??

**Comment by mprevot** [comment]
@krytarowski Can you develop in which ways 'FreeBSD support' can be better ?

**Comment by wfurt** [comment]
It would be great if some FreeBSD guru could take look at https://github.com/dotnet/coreclr/issues/22124 I would expect binaries build for 11 to run on 12 but it does not seems to be the case ;(
It is easy to reproduce with simple app and 12.0 release seems to break something dotnet depends on.

**Comment by RussellHaley** [comment]
Hi, I'm no guru but we ran across a threading regression in 12-RELEASE in the Lua53 port. 
See here for the original bug: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=235158 
and here for the Base system bug that was identified: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=235211

The fix for Lua is to link against -pthread, even though there is ZERO requirement from Lua for -pthread.

**Comment by wfurt** [comment]
thanks @RussellHaley. That looks like promising lead.

**Comment by RussellHaley** [comment]
Glad I could help. I'd love to be playing along but I barely have the few hours needed to maintain the Lua port. Keep up the great work!

**Comment by josteink** [comment]
As the one who [fixed the FreeBSD threading implementation in coreclr](https://github.com/dotnet/coreclr/commit/7976ae89dafa9be15d30210d2c56705e524c9d8b), I'm pretty sure pthreads are used pretty consistently all over, because that was what I had to patch up to make the build run.

That said, there might be underlying buts and pieces that I never had to touch which uses "regular" threads...

Maybe someone else who knows more about the general implementation can chime in? @janvorli ?

**Comment by janvorli** [comment]
Sure, please use my github user name at microsoft dot com

**Comment by janvorli** [comment]
I am sorry, I've sent the previous message in a wrong issue.

**Comment by wfurt** [comment]
This fixes the issue for me:
```
[furt@daemon ~]$ LD_PRELOAD=/usr/lib/libpthread.so ./dotnet-3.x/dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview-010021
 Commit:    d5c97b7c2a

Runtime Environment:
 OS Name:     FreeBSD
 OS Version:  12
 OS Platform: FreeBSD
 RID:         freebsd-x64
 Base Path:   /usr/home/furt/dotnet-3.x/sdk/3.0.100-preview-010021/

Host (useful for support):
  Version: 3.0.0-preview-27218-01
  Commit:  d40b87f29d

.NET Core SDKs installed:
  3.0.100-preview-010021 [/usr/home/furt/dotnet-3.x/sdk]

.NET Core runtimes installed:
  Microsoft.NETCore.App 3.0.0-preview-27218-01 [/usr/home/furt/dotnet-3.x/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

I did not make any extensive tests but at least I can execute `dotnet` again.

**Comment by janvorli** [comment]
Ok, I can see that the dotnet executable is not linked with pthreads for other systems than Linux.
https://github.com/dotnet/core-setup/blob/2ef0b64810530961f492c33d37fc7509128e0a9b/src/corehost/cli/exe.cmake#L59-L61

**Comment by josteink** [comment]
Does that mean the answer to fixing this is as simple as it sounds? I.e. as simple as this? https://github.com/josteink/core-setup/commit/25657ba2e181cce401acd7f4bf9d27a08a668470

If so, I'll be happy to make it a PR.

**Comment by wfurt** [comment]
Yes I think so. I was waiting for @joperator to confirm. 
Not sure if we need "dl" as well but that can be resolved when you submit PR @josteink

**Comment by josteink** [comment]
Right. My bad. So more like this then: https://github.com/josteink/core-setup/commit/a08f38e25a98c25f59c8ed8c8567a0cb08b1c1c6

I've created a PR for it. Let me know if it needs any ammending: https://github.com/dotnet/core-setup/pull/5072

**Comment by RussellHaley** [comment]
> Right. My bad. So more like this then: [josteink/core-setup@a08f38e](https://github.com/josteink/core-setup/commit/a08f38e25a98c25f59c8ed8c8567a0cb08b1c1c6)
> 
> I've created a PR for it. Let me know if it needs any ammending: [dotnet/core-setup#5072](https://github.com/dotnet/core-setup/pull/5072)

Just an FYI, it seems this is already patched in the base system: https://reviews.freebsd.org/D18988

**Comment by joperator** [comment]
Looks like the main issue in [dotnet/coreclr#22124](https://github.com/dotnet/coreclr/issues/22124#issuecomment-460190199) is fixed @wfurt.
I only have a problem when trying to publish a self-contained app on FreeBSD 12.0.

**Comment by hjc4869** [comment]
freebsd-x64 official NuGet packages has been removed since .NET Core 3.0 preview 2 and we were unable to publish apps for FreeBSD since then. Are there any plans to re-enable them in 3.0?

**Comment by karelz** [comment]
Sadly, we had to de-prioritize FreeBSD bring up (due to various reasons and difficulties in the end-to-end Azure support) and it is not going to be priority in .NET Core 3.0.
We would love to keep it semi-working at the state where it is now, but we do not have much time to invest now :(.

**Comment by hjc4869** [comment]
@karelz Thanks for your reply and I understand the prioritized work of .NET Core 3.0. I'll focus on bringing up my apps with FreeBSD Linux emulation then. :)

**Comment by newsash** [comment]
@hjc4869 Or you can try mono. IIRC, it will support .NET Standard 3.0

**Comment by wfurt** [comment]
I'm planning to give it another try but as @karelz mentioned it is not priority for 3.0

**Comment by hjc4869** [comment]
@newsash Mono is an acceptable option for me. However I found it difficult to build my project with the mono targets added to existing .NET Core csproj files.

On a Linux machine I tried adding net472 to TargetFrameworks and setting the FrameworkPathOverride variable but that did not work well. If I consume an API that is implemented in both mono and .NET Core, but not .NET Framework, it will fail to build with mono. Besides, although mono supports .NET Standard 2.1, I still couldn't add reference to .NET Standard 2.1 dlls in an net472 csproj.

Should I add a separate csproj and use mono msbuild instead of using dotnet tools, or do you have any suggestions on the problem?

**Comment by nkosi23** [comment]
Just a quick remark.

With mono about to be swallowed by .NET 5 as per the recent announcement [1], providing decent support for FreeBSD has become urgent.

Mono has proven to have really good cross-platform support and can be built without problem from FreeBSD ports. Many shops run their .net loads on FreeBSD since this OS has many unique features. So far, mono has been bridging the gap but with .NET 5 it seems likely that in some near future, mono will be merged into  NET 5 and the FreeBSD community will be totally cut-off from the  .NET ecosystem.

Dotnet should have mature FreeBSD support well before this happens.

I think Microsoft should officially support FreeBSD at this point and ensure all the dotnet tooling builds on this platform.

[1]: https://devblogs.microsoft.com/dotnet/introducing-net-5/

**Comment by wfurt** [comment]
@jasonpugsley put together instructions https://github.com/jasonpugsley/core-sdk/wiki/.Net-Core-3.0.0-for-FreeBSD and @joperator is trying to get source-build working https://github.com/dotnet/source-build/issues/1139

We have last ~30 tests failing for corefx. 
```
System.Diagnostics.Tests.ProcessTests.TestPeakWorkingSet64
System.Diagnostics.Tests.ProcessTests.TestPrivateMemorySize
System.Diagnostics.Tests.ProcessTests.Kill_ExitedNonChildProcess_DoesNotThrow(killTree: True)
System.Diagnostics.Tests.ProcessTests.TotalProcessorTime_PerformLoop_TotalProcessorTimeValid
System.Diagnostics.Tests.ProcessTests.Kill_EntireProcessTree_True_EntireTreeTerminated
System.Diagnostics.Tests.ProcessTests.TestPeakVirtualMemorySize
System.Diagnostics.Tests.ProcessTests.ProcessNameMatchesScriptName
System.Diagnostics.Tests.ProcessTests.TestPrivateMemorySize64
System.Diagnostics.Tests.ProcessTests.LongProcessNamesAreSupported
System.Diagnostics.Tests.ProcessTests.TestPeakWorkingSet
System.Diagnostics.Tests.ProcessTests.TestPeakVirtualMemorySize64
System.Diagnostics.Tests.ProcessTests.Kill_ExitedChildProcess_DoesNotThrow(killTree: True)
System.Diagnostics.Tests.ProcessTests.Kill_EntireProcessTree_True_CalledOnTreeContainingCallingProcess_ThrowsInvalidOperationException
System.IO.Tests.DirectoryInfo_MoveTo.MoveDirectory_FailToMoveLowerCaseDirectoryWhenUpperCaseDirectoryExists
System.IO.Tests.FileInfo_Exists.LockedFileExists
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 0, firstLength: 10, secondPosition: 1, secondLength: 2)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 3, secondLength: 5)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 3, secondLength: 4)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 4, secondLength: 5)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 2, secondLength: 6)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 2, secondLength: 4)
System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 4, secondLength: 6)
System.IO.Tests.Directory_Move.MoveDirectory_FailToMoveLowerCaseDirectoryWhenUpperCaseDirectoryExists
System.Net.NameResolution.Tests.GetHostEntryTest.Dns_GetHostEntry_HostString_Ok(hostName: \&quot;\&quot;)
System.Net.NameResolution.Tests.GetHostEntryTest.Dns_GetHostEntryAsync_HostString_Ok(hostName: \&quot;\&quot;)
System.Net.NameResolution.Tests.GetHostByNameTest.DnsObsoleteBeginEndGetHostByName_EmptyString_ReturnsHostName
System.Net.NameResolution.Tests.GetHostByNameTest.DnsObsoleteGetHostByName_EmptyString_ReturnsHostName
System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddressAndTimeoutAndBufferAndPingOptions_Unix(addressFamily: InterNetwork)
System.Net.NetworkInformation.Tests.PingTest.SendPingWithIPAddressAndTimeoutAndBufferAndPingOptions_Unix(addressFamily: InterNetwork)
System.Net.Sockets.Tests.DualModeAcceptAsync.AcceptAsyncV4BoundToSpecificV4_Success
System.Tests.AppDomainTests.MonitoringIsEnabled
System.Tests.GCExtendedTests.GetGCMemoryInfo
```
@am11 is looking at System.Diagnostics.Tests.ProcessTests so failing locking tests seems biggest remaining gap. It would be great if anybody can take look #30899.
